Move scalar_to_tensor_default_dtype out of ScalarOps.h because it's only useful for torch.where.

gchanan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #50115 Stop moving scalars to GPU for one computation in leaky_rrelu_backward.
* #50114 Stop using an unnecessary scalar_to_tensor(..., device) call.
* **#50111 Move scalar_to_tensor_default_dtype out of ScalarOps.h because it's only useful for torch.where.**
* #50105 Stop using c10::scalar_to_tensor in float_power.

Differential Revision: [D25789638](https://our.internmc.facebook.com/intern/diff/D25789638)

gchanan committed:
Move scalar_to_tensor_default_dtype out of ScalarOps.h because it's only useful for torch.where.

gchanan committed:
Update on "Move scalar_to_tensor_default_dtype out of ScalarOps.h because it's only useful for torch.where."


Differential Revision: [D25789638](https://our.internmc.facebook.com/intern/diff/D25789638)

gchanan committed:
Update on "Move scalar_to_tensor_default_dtype out of ScalarOps.h because it's only useful for torch.where."


Differential Revision: [D25789638](https://our.internmc.facebook.com/intern/diff/D25789638)

gchanan committed:
Update on "Move scalar_to_tensor_default_dtype out of ScalarOps.h because it's only useful for torch.where."


Differential Revision: [D25789638](https://our.internmc.facebook.com/intern/diff/D25789638)

gchanan committed:
Update on "Move scalar_to_tensor_default_dtype out of ScalarOps.h because it's only useful for torch.where."


Differential Revision: [D25789638](https://our.internmc.facebook.com/intern/diff/D25789638)

gchanan committed:
Update on "Move scalar_to_tensor_default_dtype out of ScalarOps.h because it's only useful for torch.where."


Differential Revision: [D25789638](https://our.internmc.facebook.com/intern/diff/D25789638)

mruberry commented on aten/src/ATen/native/TensorCompare.cpp:
```diff
@@ -195,7 +195,29 @@ bool is_nonzero(const Tensor& self) {
 namespace {
 
-static Tensor wrapped_scalar_tensor(
+at::Tensor scalar_to_tensor_default_dtype(
```
This function must still deserve some kind of comment, right? Even if it's "DO NOT USE THIS".

gchanan replied:
oh, I didn't do that because it's just an implementation detail of wrapped_scalar_tensor below -- I'll make a note of that.

mruberry commented on aten/src/ATen/native/TensorCompare.cpp:
```diff
@@ -195,7 +195,29 @@ bool is_nonzero(const Tensor& self) {
+    return at::scalar_tensor(
+        s, at::device(device).dtype(at::get_default_complex_dtype()));
+  } else {
+    AT_ASSERT(s.isIntegral(false));
```
AT_ASSERT -> TORCH_INTERNAL_ASSERT

mruberry commented on aten/src/ATen/native/TensorCompare.cpp:
```diff
@@ -195,7 +195,29 @@ bool is_nonzero(const Tensor& self) {
+  }
+}
+
+// `use_default_dtype` is a bit of a hack because torch.where doesn't support type promotion, but
```
Wow this comment really dives in!

Is this function only intended to be used by torch.where? A sentence upfront about its intended use would be helpful. In particular, if torch.where supported type promotion properly would you expect us to get rid of this function, and should this comment tell developers not to use this function?

gchanan replied:
ya, I can add a sentence upfront summarizing.

gchanan committed:
Update on "Move scalar_to_tensor_default_dtype out of ScalarOps.h because it's only useful for torch.where."


Differential Revision: [D25789638](https://our.internmc.facebook.com/intern/diff/D25789638)

mruberry reviewed (approved):
Cool
